File: credentials-verification-android/app/src/main/java/io/iohk/cvp/viewmodel/NewConnectionsViewModel.java
package io.iohk.cvp.viewmodel;

import android.os.AsyncTask;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import javax.inject.Inject;

import io.iohk.cvp.grpc.AsyncTaskResult;
import io.iohk.cvp.grpc.GetConnectionTokenInfoRunnable;
import io.iohk.cvp.grpc.GetMessagesRunnable;
import io.iohk.cvp.grpc.GrpcTask;
import io.iohk.cvp.grpc.ParticipantInfoResponse;
import io.iohk.prism.protos.ParticipantInfo;

public class NewConnectionsViewModel extends CvpViewModel {

    private MutableLiveData<AsyncTaskResult<ParticipantInfoResponse>> issuerInfo = new MutableLiveData<>();

    @Inject
    public NewConnectionsViewModel() {
    }

    public LiveData<AsyncTaskResult<ParticipantInfoResponse>> getConnectionTokenInfoLiveData() {
        return issuerInfo;
    }

    public void getConnectionTokenInfo(String token) {
        GrpcTask task = new GrpcTask<>(new GetConnectionTokenInfoRunnable(issuerInfo), context);
        task.execute(null,token);
        runningTasks.add(task);
    }
}

File: credentials-verification-android/app/src/main/java/io/iohk/cvp/viewmodel/NewConnectionsViewModel.kt
package io.iohk.cvp.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import io.iohk.cvp.data.DataManager
import io.iohk.cvp.data.local.db.model.Contact
import io.iohk.cvp.grpc.AsyncTaskResult
import io.iohk.cvp.grpc.ParticipantInfoResponse
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

open class NewConnectionsViewModel() : CvpViewModel() {

    private lateinit var appDataManager: DataManager

    private val issuerInfo = MutableLiveData<AsyncTaskResult<ParticipantInfoResponse>>()

    constructor(dataManager: DataManager) : this() {
        appDataManager = dataManager
    }

    fun getConnectionTokenInfoLiveData(): LiveData<AsyncTaskResult<ParticipantInfoResponse>>? {
        return issuerInfo
    }

    fun getConnectionTokenInfo(token: String) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val participantInfoResponse = appDataManager.getConnectionTokenInfo(token)
                val connectionsList = appDataManager.getAllContacts()
                val found = connectionsList.any {
                    connection: Contact ->
                    connection.name == participantInfoResponse.creator.issuer.name || connection.name == participantInfoResponse.creator.holder.name
                }
                issuerInfo.postValue(AsyncTaskResult(ParticipantInfoResponse(participantInfoResponse.creator, token, found)))
            } catch (ex:Exception) {
                issuerInfo.postValue(AsyncTaskResult(ex))
            }
        }
    }
}
